### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.trainer.connectors.data_connector.py`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to `pytorch_lightning.trainer.connectors.data_connector.py` for issue #13445 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:462]: SequentialSampler is a PyTorch core method that requires arg 1 to be of type `Sized` however, `dataloader.dataset` is torch.utils.data.dataset, and is a `Generic[T_co]` type.

This appears to be an issue with PyTorch core's type hints.

File: src/pytorch_lightning/core/module.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/module.py:102]: added to resolve

```sh
src/pytorch_lightning/trainer/connectors/data_connector.py:148: error: Incompatible types in assignment (expression has type "Union[str, int]", variable has type "int")  [assignment]
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:355]: associated error ignored:

```sh
src/pytorch_lightning/trainer/connectors/data_connector.py:355: error: List item 0 has incompatible type "Union[DataLoader[Any], Collection[DataLoader[Any]]]"; expected "DataLoader[Any]"  [list-item]
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:17]: Hi @JustinGoheen, all of the tests are now failing, as none of them are running on 3.10, which is the version when `TypeAlias` became available. https://docs.python.org/3/library/typing.html#typing.TypeAlias In order to fix it, please import `TypeAlias` from `typing_extensions`

Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: updated in https://github.com/Lightning-AI/lightning/pull/13806/commits/87cb7a66d3ed03cdca5ed8ae60d211a7049e329e

Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: Well, another error which made every test broke popped up -> `TypeAlias` is not subscriptable

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:148]: I don't think `TypeAlias` is appropriate - what are you trying to solve here?

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:356]: This is not a proper way, how to resolve this. I'd rather see it resolved in methods generating this value (i.e. `self._resolve_overfit_batches` and `self._request_dataloader`)

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:500]: what is `TypeAlias["instance"]` supposed to be?

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:552]: I don't think that limiting `_valid_hooks` to length 3 is a proper approach. It makes sense to make it variable length (https://docs.python.org/3/library/typing.html#typing.Tuple)

**Review by otaj — CHANGES_REQUESTED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: not able to find the original error in CI/CD checks; I removed locally and it is fine. Updated in https://github.com/Lightning-AI/lightning/pull/13806/commits/82ee684a0b76486eee08b0b9f7462762edf3ae51

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:500]: without TypeAlias I get the following error:

```sh
src/pytorch_lightning/trainer/connectors/data_connector.py:500: error: Variable "pytorch_lightning.trainer.connectors.data_connector._DataLoaderSource.instance" is not valid as a type  [valid-type]
src/pytorch_lightning/trainer/connectors/data_connector.py:500: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: updated in https://github.com/Lightning-AI/lightning/pull/13806/commits/6020d0b3ec9afa56ee34757adea889163eac16bc

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: `_resolve_overfit_batches` is [used in trainer](https://github.com/Lightning-AI/lightning/blob/4e9fbdbe4013458ea33256f4e0b00cf0782e6064/src/pytorch_lightning/trainer/trainer.py#L1813) in a way that does not require a list be returned; updating `_resolve_overfit_batches` to fix the error here, would require that I alter code in trainer.

I tried `dataloaders = list(dataloaders)` out of curiosity, and that worked locally.

Updated in https://github.com/Lightning-AI/lightning/pull/13603/commits/ca4f1609731b64201c3fe184fe71a104fd15276d

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:500]: The issue here seems that both of the types (`self.instance` as well as `self.dataloader()`) are valid, but mypy has a bit of a trouble here, since it thinks that it could happen, that `self.instance` is `pl.LightningModule` or `pl.LightningDataModule` and we return that value. Which is not true, as evidenced by the method, but mypy just doesn't understand it. In this case, I don't think there's an easy way to fix this and I'd go with `# type: ignore`

Any other opinions, @awaelchli @carmocca?

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: But `list(dataloaders) != [dataloaders]`. `list(iterable)` tries to convert that iterable into list (which would succeed for single dataloader, but is definitely not something we want here), whereas `[iterable]` creates a list with `iterable` as a first element. No wonder mypy stopped complaining, other tests had to start complaining.

Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: Anyway, after taking a look that changing the methods is probably not feasible within this PR, I think in this case it makes sense to again, `# type: ignore`

Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: Sorry about this

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: I created a constant `INSTANCE = Optional[Union[TRAIN_DATALOADERS, EVAL_DATALOADERS, "pl.LightningModule", "pl.LightningDataModule"]]` and updated the return to `Union[INSTANCE, TRAIN_DATALOADERS, EVAL_DATALOADERS]`; this works locally, updated in https://github.com/Lightning-AI/lightning/pull/13806/commits/1aebec4fe5f7cb291784e2164367c60a578b8f79

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:356]: reverted in https://github.com/Lightning-AI/lightning/pull/13806/commits/c4039b1bc929aeb96049cfbd85e3b032f2166f14

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: to recreate the original error, I removed the return type in https://github.com/Lightning-AI/lightning/pull/13806/commits/b0cb0409a05469990f4805b953ed937f06707bd3

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: TypeAlias has been removed

File: src/pytorch_lightning/strategies/deepspeed.py
Comment (JustinGoheen) [inline src/pytorch_lightning/strategies/deepspeed.py:677]: Once I added Instance as a return type in `data_connector._DataLoaderSource.dataloader`, this error was raised by mypy.

```sh
src/pytorch_lightning/strategies/deepspeed.py:677: error: Incompatible types in assignment (expression has type "Union[int, Any, Tensor, Module]", variable has type "int")  [assignment]
```

I traced back to `data_connector._DataLoaderSource.dataloader`, and there is no batch_sampler or batch_size in that specific method, or in the entire module.

`batch_sampler` and `batch_size` are attributes of torch.utils.data.DataLoader; given this is outside of the scope of PL, I opted to ignore the error.

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: reverted in https://github.com/Lightning-AI/lightning/pull/13806/commits/dfca1fe56087253e8fc3fb2bab52b7bdcce72377

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:500]: The original types were correct, this method cannot return `LightningModule` or `LightningDataModule`. Either resolve it in the method (which I'm not really sure how feasible it is, since that method does a bit of magic, that `mypy` might not like) or add proper `# type: ignore`s, but don't extend the result type of this method

Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: I opted to ignore after attempting to use self.is_defined to block mypy from evaluating return-value as a LightningModule.

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:50]: Last nit: I'd rename this to something more descriptive

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: it's only used once now; so I just moved the annotation back to the appropriate class attr in https://github.com/Lightning-AI/lightning/pull/13806/commits/c2431aeb555bace222bbe470d8a60c381458f125

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:164]: I think adding an entry here:
https://github.com/Lightning-AI/lightning/blob/ee4233b36cb70b19a13013371ada50875b429ef3/src/pytorch_lightning/trainer/trainer.py#L634
```py
self.predict_dataloaders = None
```

will help avoid `ignore[assignment]` here.

File: src/pytorch_lightning/trainer/trainer.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:2243]:
```suggestion
    def distributed_sampler_kwargs(self) -> Optional[Dict[str, int]]:
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:507]: do you know why we need to ignore here?

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:424]:
```suggestion
    def _request_dataloader(self, stage: RunningStage) -> TRAIN_DATALOADERS:
```
from `utilities/types.py`

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:443]:
```suggestion
    def _resolve_overfit_batches(dataloaders: EVAL_DATALOADERS, mode: RunningStage) -> EVAL_DATALOADERS:
```
from `utilities/types.py`

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:50]:
```suggestion
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:424]: this suggestion causes several errors. the new type was made to address specific errors encountered.

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:507]: https://github.com/Lightning-AI/lightning/pull/13806#discussion_r935491977

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:443]: combining these suggestions with the other suggestions also causes additional errors

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:50]: these are needed to resolve errors. each was created based on mypy error output.

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:164]: doing this causes new errors 

```sh
src/pytorch_lightning/loops/dataloader/prediction_loop.py:63: error: Incompatible return value type (got "None", expected "Sequence[DataLoader[Any]]")  [return-value]
```

above error can be resolved with

```python
def dataloaders(self) -> Sequence[DataLoader]:
    """Returns all prediction dataloaders."""
    if self.trainer.predict_dataloaders:
        return self.trainer.predict_dataloaders
```

is it appropriate to annotate in Trainer, and update logic in a third file in order to remove an ignore?

File: src/pytorch_lightning/trainer/trainer.py
Comment (JustinGoheen) [reply src/pytorch_lightning/trainer/trainer.py:2243]: this causes

```sh
src/pytorch_lightning/trainer/connectors/data_connector.py:296: error: Argument after ** must be a mapping, not "Optional[Dict[str, int]]"  [arg-type]
```

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: added a commit. removed 4 `ignores`

**Review by rohitgr7 — APPROVED** [review]

**Comment by rohitgr7** [comment]
messed up somewhere :p
will check tomorrow!

**Comment by JustinGoheen** [comment]
apologies, I'm no longer able to keep this PR up to date.

**Comment by otaj** [comment]
@JustinGoheen, we're sorry to hear that. @rohitgr7, are you gonna push it over the finish line or should I? I noticed you started pushing some commits :zap:

**Comment by rohitgr7** [comment]
@otaj it's done. Any things else needs to be done here? The CI is blocking the merge for now

**Comment by otaj** [comment]
@rohitgr7, awesome, thank you! :tada:. I guess we just have to wait on codeowners and CI

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:504]: Is this removed on purpose?

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:445]: Isn't this also used with training dataloaders?

File: src/pytorch_lightning/trainer/trainer.py
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:2234]: This would still be Optional

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:504]: this is covered by the default return at the end if `self.instance` is neither LM nor LDM.

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:445]: good catch! let me update

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:295]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: I agree, but I can't find them anymore :joy: Marking as resolved so we can merge this

Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: looks like an old review before new updates

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**
